storage/replica_raftstorage.go

```diff
@@ -484,24 +484,11 @@ func (r *Replica) applySnapshot(batch engine.Engine, snap raftpb.Snapshot) (uint
 
 	rangeID := r.RangeID
 
-	// First, save the HardState. The HardState must not be changed
-	// because it may record a previous vote cast by this node. This is
-	// usually unnecessary because a snapshot is nearly always
-	// accompanied by a new HardState which incorporates both our former
-	// state and new information from the leader, but in the event that
-	// the HardState has not changed, we want to use our own previous
-	// HardState and not one that was transmitted via the snapshot.
-	hardStateKey := keys.RaftHardStateKey(rangeID)
-	hardState, _, err := engine.MVCCGet(batch, hardStateKey, roachpb.ZeroTimestamp, true /* consistent */, nil)
-	if err != nil {
-		return 0, err
-	}
-
 	// Extract the updated range descriptor.
 	desc := snapData.RangeDescriptor
 
 	// Delete everything in the range and recreate it from the snapshot.
-	iter := newReplicaDataIterator(&desc, batch)
+	iter := newReplicaDataIterator(&desc, batch, true /* replicatedOnly */)
```

Comment:
We need to delete any old log entries here: any log entries that predate the snapshot will be orphaned and never truncated or GC'd. It's probably better to continue to treat the HardState as a special case; it should be the only thing that needs to be preserved across `applySnapshot`; the other unreplicated state can be discarded.